File: 0956-30-2185-counting-words-with-a-given-prefix/0956-30-2185-counting-words-with-a-given-prefix.py
class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.frq = 0

class Solution:
    def __init__(self):
        self.trie = TrieNode()

    def insert(self, word):
        node = self.trie
        for ch in word:
            index = ord(ch) - ord('a')
            if not node.children[index]:
                node.children[index] = TrieNode()
            node = node.children[index]
            node.frq += 1

    def search(self, prefix):
        node = self.trie
        for ch in prefix:
            index = ord(ch) - ord('a')
            if not node.children[index]:
                return 0
            node = node.children[index]
        return node.frq

    def prefixCount(self, words, pref):
        for word in words:
            self.insert(word)
        return self.search(pref)

File: 0956-30-2185-counting-words-with-a-given-prefix/0956-30-2185-counting-words-with-a-given-prefix.java
class TrieNode {
    TrieNode[] children = new TrieNode[26];
    int frq = 0;
}

class Solution {
    private TrieNode trie;

    public Solution() {
        this.trie = new TrieNode();
    }

    private void insert(String word) {
        TrieNode node = trie;
        for (char ch : word.toCharArray()) {
            int index = ch - 'a';
            if (node.children[index] == null) {
                node.children[index] = new TrieNode();
            }
            node = node.children[index];
            node.frq++;
        }
    }

    private int search(String prefix) {
        TrieNode node = trie;
        for (char ch : prefix.toCharArray()) {
            int index = ch - 'a';
            if (node.children[index] == null) {
                return 0;
            }
            node = node.children[index];
        }
        return node.frq;
    }

    public int prefixCount(String[] words, String pref) {
        for (String word : words) {
            insert(word);
        }
        return search(pref);
    }
}
